omniledger/contracts/coins.go

```diff
@@ -123,35 +124,37 @@ func ContractCoin(cdb ol.CollectionView, inst ol.Instruction, c []ol.Coin) (sc [
 				return
 			}
 
-			targetCoin := newSafeUint64(v)
-			targetCoin, err = targetCoin.add(coinsArg)
+			var targetCI CoinInstance
+			err = protobuf.Decode(v, &targetCI)
+			if err != nil {
+				return nil, nil, errors.New("couldn't unmarshal target account: " + err.Error())
+			}
+			targetCI.Balance, err = safeAdd(targetCI.Balance, coinsArg)
 			if err != nil {
 				return
 			}
-			var w bytes.Buffer
-			binary.Write(&w, binary.LittleEndian, targetCoin)
+			targetBuf, err := protobuf.Encode(&targetCI)
+			if err != nil {
+				return nil, nil, errors.New("couldn't marshal target account: " + err.Error())
+			}
 
 			log.Lvlf3("transferring %d to %x", coinsArg, target)
 			sc = append(sc, ol.NewStateChange(ol.Update, ol.NewInstanceID(target),
-				ContractCoinID, w.Bytes(), did))
+				ContractCoinID, targetBuf, did))
 		case "fetch":
 			// fetch removes coins from the account and passes it on to the next
 			// instruction.
-			if coinsArg > uint64(coinsCurrent) {
-				err = errors.New("not enough coins in instance")
-				return
-			}
-			coinsCurrent, err = coinsCurrent.sub(coinsArg)
+			ci.Balance, err = safeSub(ci.Balance, coinsArg)
 			if err != nil {
 				return
 			}
-			cOut = append(cOut, ol.Coin{Name: CoinName, Value: coinsArg})
+			cOut = append(cOut, ol.Coin{Name: ol.NewInstanceID(ci.Type), Value: coinsArg})
 		case "store":
 			// store moves all coins from this instruction into the account.
 			cOut = []ol.Coin{}
 			for _, co := range c {
 				if bytes.Equal(co.Name.Slice(), CoinName.Slice()) {
```

Comment:
Fetch is no longer using CoinName, but store is?

Reply:
Nice catch.